src/Data/Singletons/Promote/Type.hs

```diff
@@ -7,26 +7,37 @@ This file implements promotion of types into kinds.
 -}
 
 module Data.Singletons.Promote.Type
-  ( promoteType, promoteTypeArg, promoteUnraveled
+  ( promoteType, promoteType_NC
+  , promoteTypeArg_NC, promoteUnraveled
   ) where
 
 import Language.Haskell.TH.Desugar
 import Data.Singletons.Names
+import Data.Singletons.Util
 import Language.Haskell.TH
 
--- the only monadic thing we do here is fail. This allows the function
--- to be used from the Singletons module
+-- Promote a DType to the kind level.
+--
+-- NB: the only monadic thing we do here is fail. This allows the function
+-- to be used from the Singletons module.
 promoteType :: MonadFail m => DType -> m DKind
-promoteType = go []
+promoteType ty = do
+  checkVanillaDType ty
+  promoteType_NC ty
+
+-- Promote a DType to the kind level. This is suffixed with "_NC" because
+-- we do not invoke checkVanillaDType here.
+-- See [Vanilla-type validity checking during promotion].
+promoteType_NC :: MonadFail m => DType -> m DKind
+promoteType_NC = go []
   where
     go :: MonadFail m => [DTypeArg] -> DType -> m DKind
+    go []       (DForallT _fvf _tvbs ty) = go [] ty
```

Comment:
The comment explains why we drop the constraint. Good. But why drop the `forall`?

Reply:
Good question. Now that we've separated out `DForallT` and `DConstrainedT`, I don't think there's any reason to drop explicit `DForallT`s any more... at least in theory. In practice, I think this code path is unused at the moment, since if I apply this change:

```diff
diff --git a/src/Data/Singletons/Promote/Type.hs b/src/Data/Singletons/Promote/Type.hs
index d35ac75..d9f893d 100644
--- a/src/Data/Singletons/Promote/Type.hs
+++ b/src/Data/Singletons/Promote/Type.hs
@@ -32,7 +32,9 @@ promoteType_NC :: MonadFail m => DType -> m DKind
 promoteType_NC = go []
   where
     go :: MonadFail m => [DTypeArg] -> DType -> m DKind
-    go []       (DForallT _fvf _tvbs ty) = go [] ty
+    go []       (DForallT fvf tvbs ty) = do
+      ty' <- go [] ty
+      pure $ DForallT fvf tvbs ty'
     -- We don't need to worry about constraints: they are used to express
     -- static guarantees at runtime. But, because we don't need to do
     -- anything special to keep static guarantees at compile time, we don't
```

Then the test suite continues to pass without needing any changes whatsoever. As luck would have it, `promoteType` only ever seems to be called on types without top-level `forall`s. For example, `D.S.Single.Type.singType`, which singles top-level type signatures, splits the type into argument types (discarding top-level `forall`s) and then invokes `promoteType` on each argument type individually.

I could foresee a future in which preserving `DForallT` in `promoteType` is important, however. If we implement #378, for instance, we'll need to take this function:

```hs
blah :: forall b a. [a] -> [b -> a]
blah = ...
```

And promote it to this:

```hs
type Blah :: forall b a. [a] -> [b ~> a]
type family Blah (x :: [a]) :: [b ~> a] where ...
```

The standalone kind signature for `Blah` will be the output of `promoteType`, so in order to ensure that `forall b a.` appears in that SAK (and in exactly that order, to preserve its properties _vis-à-vis_ `TypeApplications`), `promoteType` needs to preserve `forall`s. For this reason, I've gone ahead and just applied the change above.

Reply:
Sounds good. :)